refactor: Scheduling based on KV cache state

Funatiq opened the pull request:
## Description

Use `mSequences.at(req.mRequestId).getNumTokens()` instead of `req.getMaxNumGeneratedTokens()` to determine scheduling. This allows to disable two-step lookahead in overlap scheduling because the former is updated at the beginning of each iteration while the latter is only updated at the end.

See commit messages for details.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
refactor: Update token management in KVCacheManager for improved block calculation

- Enhanced the getNeededBlocksOneStep method to utilize current token counts for more accurate block requirements.
- Adjusted the logic to account for the number of tokens already processed, improving efficiency in scheduling.
- Don't use two-step lookahead in CapacityScheduler for overlap scheduling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Simplify block calculation logic in KVCacheManager

- Streamlined the getNeededBlocksOneStep method by consolidating token management and block calculation.
- Removed redundant variable declarations and improved readability by directly returning calculated values.
- Enhanced handling of token limits for more efficient scheduling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #7337](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7337/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7337](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7337/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5319](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5319/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7366](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7366/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7464](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7464/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7464](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7464/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5418](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5418/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #7923](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7923/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7923](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7923/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #5728](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5728/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #8468](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8468/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6133 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6133/) completed with status: 'SUCCESS'

Funatiq committed:
refactor: Move NoEvictScheduledBlocksManager and MaxUtilizationScheduledBlocksManager classes to separate header file

- Moved NoEvictScheduledBlocksManager and MaxUtilizationScheduledBlocksManager from kvCacheManager.h to a new scheduledBlocksManager.h.
- Updated capacityScheduler.cpp to include the new scheduledBlocksManager header.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --disable-multi-gpu-test

Funatiq commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #8695](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8695/) [ run ] triggered by Bot

Funatiq commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1631,46 +1631,48 @@ void KVCacheManager::startScheduling()
     }
-    return numRequiredBlocks;
+
+    return 0;
```
Yes, if context chunking is in progress or the request is already finished.

Funatiq commented on cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:
```diff
@@ -398,7 +398,8 @@ TrtGptModelInflightBatching::TrtGptModelInflightBatching(std::shared_ptr<nvinfer
     mCapacityScheduler = std::make_unique<CapacityScheduler>(getMaxNumSequences(),
-        optionalParams.schedulerConfig.getCapacitySchedulerPolicy(), mKvCacheManager != nullptr, mNumMicroBatches > 1);
+        optionalParams.schedulerConfig.getCapacitySchedulerPolicy(), mKvCacheManager != nullptr,
+        mWorldConfig.isPipelineParallel());
```
No, `mNumMicroBatches > 1` is true for overlap and PP, but we only want to cover the PP case.

tensorrt-cicd commented:
[PR_Github #8695](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8695/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6303 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6303/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #8749](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8749/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8749](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8749/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6350 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6350/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/6350/test-results/rerun_report.html) for details.

Funatiq commented:
/bot skip --comment "Pipelines 8626 + 8749 have passed all tests"

tensorrt-cicd commented:
[PR_Github #8797](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8797/) [ skip ] completed with state `SUCCESS`
Skipping testing for commit `f42bdc2`

MartinMarciniszyn commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1633,46 +1633,48 @@ void KVCacheManager::startScheduling()
             = tc::ceilDiv(numUnSharedTokens, getTokensPerBlock()) * req.mSamplingConfig.beamWidth;
-        numRequiredBlocks = numSharedBlocks + numUnSharedBlocks;
+        auto const numRequiredBlocks = numSharedBlocks + numUnSharedBlocks;
+        return numRequiredBlocks;
```
Maybe inline?

MartinMarciniszyn reviewed (approved):
Approved.
